Plugin to prevent special characters in code blocks

frozenprocess opened the pull request:
## Description
@lxpollitt @joshti
This plugin search for special characters in code blocks and prevents compilation cycle from completion with an error message to user.

I intentionally did not modify `networking/bgp.md` and `felix.md` file in order to break Semaphore compile cycle in first compilation attempt, if you like to see the error message created by this plugin please check the first build attempt in Semaphore.


## Related issues/PRs
#3769 Original place that this problem was mentioned.

```release-note
None required
```

caseydavenport commented:
Deploy preview for *calico* ready!

Built with commit 929b7212260f6721325ef07207df7757be248dbe

https://deploy-preview-3775--calico.netlify.app

frozenprocess commented:
Error example generated by plugin
```
Incremental build: enabled
      Generating... 
Line: calicoctl patch node node-1 -p '{"spec": {"bgp": {“asNumber”: “64514”}}}'
                                                        ^~~~~~~~~~~~
Line: calicoctl patch node node-1 -p '{"spec": {"bgp": {“asNumber”: “64514”}}}'
                                                                 ^~~~~~~~~~~~
Line: calicoctl patch node node-1 -p '{"spec": {"bgp": {“asNumber”: “64514”}}}'
                                                                    ^~~~~~~~~~~~
Line: calicoctl patch node node-1 -p '{"spec": {"bgp": {“asNumber”: “64514”}}}'
                                                                          ^~~~~~~~~~~~
jekyll 3.8.7 | Error:  
 Unexpected character in codeblock detected: networking/bgp.md
```

joshti commented on _plugins/code_parser.rb:
````diff
@@ -0,0 +1,20 @@
+# Description : This plugin checks code blocks created using ``` for ASCCI characters that can cause problem pasted to terminal.
````
UTF-8* Not ASCII.

ASCII is usable by kubectl / vast majority of cross-platform tooling. The universal UTF-8 conversion is well underway but it's been in progress for 20 years 😄

frozenprocess replied:
I started this plugin by converting everything to ASCII and checking for its Decimal representation!
I have changed everything but seems like that miss spelled ASCII was extremely lucky :D

joshti commented on _plugins/code_parser.rb:
```diff
@@ -0,0 +1,20 @@
+    # Check content of the post for code blocks.
+    code_blocks = post.content.scan(/<code.*>(.*?)<\/code>/m)
+    # Check code blocks for characters and saving their position in order to generate log.
+    occurences = code_blocks.join(",").gsub(/[–—‘’‚“”„…]/).map { Regexp.last_match.begin(0) }
```
I haven't touched Ruby since 1.8 (so forgive me a little as it's been well over a decade 😄).

I think we may be able to do something a little more clever here than matching the existing known problematic utf-8 characters. With Ruby 1.9 it looks like string encoding was added ( https://ruby-doc.org/core-1.9.3/String.html#method-i-encode ). What if we did  `Encoding.find('ASCII')` for the type, and then on conversion failure print the offending character in the callback. This way if the character can be converted, they will (or can) be, and if it simply cannot be converted (multiple failure modes), user intervention is required.

What are your thoughts on this one?

frozenprocess replied:
Same here, I did use Ruby while ago and its like learning it again from scratch now 😄 .

Initially I proposed something similar. however, after talking to @lxpollitt we decided instead of converting characters break the compiling cycle and ask user to intervene .

joshti commented on _plugins/code_parser.rb:
```diff
@@ -0,0 +1,20 @@
+        # Stop compiling 
+        raise "\n Unexpected character in codeblock detected: #{post.path}"
+    end
+end
```
newline please 😄

frozenprocess replied:
:D

joshti commented on networking/bgp.md:
```diff
@@ -54,13 +54,24 @@ For a deeper look at common on-premises deployment models, see [Calico over IP F
-- [View BGP peering status for a node](#view-bgp-peering-status-for-a-node)
-- [Change the default global AS number](#change-the-default-global-as-number)
-- [Change AS number for a particular node](#change-as-number-for-a-particular-node)
+- [Big picture](#big-picture)
```
It looks like another cset / stash sneaked in here.

frozenprocess replied:
At the moment this plugin only scans code blocks which are created when using one or three \`.

frozenprocess replied:
I did write another one to fully address #3769 that replace or converts these characters globally.

joshti replied:
Why did this section change though? There's nothing to do with another character set in here, menu items are completely different.

frozenprocess replied:
That was a mistake on my end instead of applying the plugin to pre_render I did it on post_render.

joshti commented on reference/legal/felix.md:
```diff
@@ -3488,28 +3488,28 @@ canonical_url: '/reference/legal/felix'
 
 		1\. Definitions
 
-		1.1. “Contributor”
```
We don't want to change anything in legal.

frozenprocess commented:
I've added my original way of character detection and reverted all old documents to their original state.

Changes:
* Plugin skips urls that contain `/legal/`.
* Documents will be examined correctly at their Markdown state.
* Debugging output can now handle multiline sentences.

@joshti WDYT?

joshti commented on _plugins/code_block_validator.rb:
```diff
@@ -0,0 +1,35 @@
+#####
+Jekyll::Hooks.register :pages, :pre_render do |post|
+    # do nothing if url indicates legal folder
+    if !post.path.match(/\/legal\//)
```
This is a pretty large branch, does ruby have return statements?

frozenprocess replied:
Ruby has returns but unfortunately it is not possible to use it in Jekyll hooks, such instruction triggers an exception.
I'll see what else can be used here.

frozenprocess replied:
The only thing that I could come up with was getting rid of line 7 `if !post.path.match(/\/legal\//)` and adding it to line 10 `post.content.scan(/(?:[`]{3}(.*?)[`]{3})/m)` transforming it to an if statement with `&&` operator to first evaluate the path and then check for regex pattern.
@joshti WTDY?

joshti replied:
Nah this makes more-sense this way.

joshti commented on _plugins/code_block_validator.rb:
```diff
@@ -0,0 +1,35 @@
+                end
+            end
+        }
+        # was there any special characters?
```
were*
